File: 0875-koko-eating-bananas/0875-koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function (piles, h) {
    let low = 1, high = Math.max(...piles);

    const canEat = (k) => {
        const time = piles.reduce(
            (acc, pile) => Math.ceil(pile / k) + acc,
            0);
        return time <= h;
    }

    while (low < high) {
        const mid = Math.floor((low + high) / 2);

        if (canEat(mid)) {
            high = mid;
        } else {
            low = mid + 1;
        }

    }

    return high;
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    private boolean canEat(int[] items, int mid, int total) {
        int hours = 0;
        for (int item : items) {
            hours += Math.ceilDiv(item, mid);
        }
        return hours <= total;
    }

    public int minEatingSpeed(int[] piles, int h) {
        int low = 1, high = 0;
        // int high = Arrays.stream(piles).max().getAsInt();
        for (int pile : piles) {
            high = Math.max(high, pile);
        }

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (canEat(piles, mid, h)) {
                high = mid;
            } else {
                low = mid + 1;
            }

        }
        return high;// low
    }
}